src/block/BaseSign.php

```diff
@@ -235,4 +234,35 @@ public function updateText(Player $author, SignText $text) : bool{
 	public function asItem() : Item{
 		return ($this->asItemCallback)();
 	}
+
+	private function getColorFromItem(Item $item) : ?DyeColor{
+		return $item instanceof Dye ? $item->getColor() : match($item->getTypeId()){
+			ItemTypeIds::BONE_MEAL => DyeColor::WHITE(),
+			ItemTypeIds::LAPIS_LAZULI => DyeColor::BLUE(),
+			ItemTypeIds::COCOA_BEANS => DyeColor::BROWN(),
+			default => null
+		};
+	}
+
+	/**
+	 * Returns whether player can open sign editor
+	 */
+	private function canOpenSignEditor(Item $item) : bool{
+		$currentColor = $this->text->getBaseColor();
+		$newColor = $this->getColorFromItem($item);
+		if($newColor === null){
+			if($item->getTypeId() === ItemTypeIds::GLOW_INK_SAC || $item->getTypeId() === ItemTypeIds::INK_SAC){
+				$wasGlowing = $this->text->isGlowing();
+				if($wasGlowing && $item->getTypeId() !== ItemTypeIds::GLOW_INK_SAC){
+					return false;
+				}
+				if(!$wasGlowing && $item->getTypeId() !== ItemTypeIds::INK_SAC){
+					return false;
+				}
+			}
+			return true;
+		}
+		$color = $newColor->equals(DyeColor::BLACK()) ? new Color(0, 0, 0) : $newColor->getRgbValue();
+		return $currentColor->equals($color);
+	}
```

Comment:
This function looks like it's repeating a lot of logic. Seems to me like we should just be able to check if the previous actions did anything in `onInteract` instead of having all this extra code.

Reply:
We can't put `openSignEditor` at the end of `onInteract()` because of the following client behavior:
* Interacting with dyes/(glow) ink sak
   * With same color/glowing state: Player can open sign editor.
   * Different color/glowing state: Sign color/glowing state will be changed.
* Interacting without any items on hand
   * Open sign editor.
  
Opening sign editor should happen before changing sign.

Reply:
I don't see how this prevents it from being placed at the end.